flow/layers/layer_tree.cc

```diff
@@ -110,24 +110,12 @@ void LayerTree::Paint(CompositorContext::ScopedFrame& frame,
     return;
   }
 
-  SkISize canvas_size = frame.canvas()->getBaseLayerSize();
-  SkNWayCanvas internal_nodes_canvas(canvas_size.width(), canvas_size.height());
-  internal_nodes_canvas.addCanvas(frame.canvas());
-  if (frame.view_embedder() != nullptr) {
-    auto overlay_canvases = frame.view_embedder()->GetCurrentCanvases();
-    for (size_t i = 0; i < overlay_canvases.size(); i++) {
-      internal_nodes_canvas.addCanvas(overlay_canvases[i]);
-    }
-  }
   DisplayListBuilder* builder = frame.display_list_builder();
-  DisplayListBuilderMultiplexer builder_multiplexer;
+  LayerStateStack state_stack;
   if (builder) {
-    builder_multiplexer.addBuilder(builder);
-    if (frame.view_embedder()) {
-      for (auto* view_builder : frame.view_embedder()->GetCurrentBuilders()) {
-        builder_multiplexer.addBuilder(view_builder);
-      }
-    }
+    state_stack.setBuilderDelegate(builder);
+  } else {
+    state_stack.setCanvasDelegate(frame.canvas());
```

Comment:
Here, don't we need to record the `frame.view_embedder` canvases?

Reply:
The canvas here is a wrapper around the builder. If we set both then the builder would get double-state-notices.